BPA/Ch.12/Nmix2.stan

```diff
@@ -1,145 +1,171 @@
-    for (k in first[i]:last[i]) {     // Loop over days
-      lp[i, k, 1:max_y[i, k]] <- rep_vector(negative_infinity(), max_y[i, k]);
-      for (n in max_y[i, k]:K)
-        lp[i, k, n + 1] <- poisson_log_log(n, log_lambda[i, k])
-          + binomial_logit_log(y[i, 1:T, k], n, logit_p[i, 1:T, k]);
-    }
-  }
+  eps = sd_lam * eps_raw;
+  for (k in 1:7)
+    for (i in 1:R)
+      log_lambda[i, k] = alpha_lam[k] + eps[i];
 }
 
 model {
   // Priors
   alpha_lam ~ normal(0, sqrt(10));
   beta ~ normal(0, sqrt(10));
   eps_raw ~ normal(0, 1);
-  // Half-Cauchy priors are used on sd_lam and sd_p, instead of
-  // uniform(0, 3) used in the book;
-  sd_lam ~ cauchy(0, 2.5);
-  sd_p ~ cauchy(0, 2.5);
 
-  for (i in 1:R)                      // Loop over R sites (95)
-    for (k in first[i]:last[i])       // Loop over days
-      logit_p[i, 1:T, k] ~ normal(beta[k], sd_p);
+  // Weakly informative priors are used on sd_lam and sd_p,
+  // instead of uniform(0, 3) used in the book
+  sd_lam ~ normal(1.5, 0.75);
+  sd_p ~ normal(1.5, 0.75);
+
+  for (i in 1:R)
+    for (j in 1:T)
+      logit_p[i, j] ~ normal(beta, sd_p);
 
   // Likelihood
-  for (i in 1:R)                      // Loop over R sites (95)
-    for (k in first[i]:last[i])       // Loop over days
-      increment_log_prob(log_sum_exp(lp[i, k, (max_y[i, k] + 1):(K + 1)]));
+  for (i in 1:R)
+    for (k in first[i]:last[i])
+      y[i, 1:2, k] ~ bivariate_poisson(log_lambda[i, k], logit_p[i, 1:2, k]);
 }
 
 generated quantities {
-  int totalN[7];
+  int totalN[7];            // Total pop. size across all sites
   vector[7] mean_abundance;
   vector[7] mean_N;
   vector[7] mean_detection;
-  real fit;
-  real fit_new;
+  real fit = 0;
+  real fit_new = 0;
 
   {
-    int N[R, 7];
-    real eval[R, T, 7];
-    real y_new[R, T, 7];
-    real p[R, T, 7];
+    int N[R, 7];          // Abundance
+    real eval[R, T, 7];   // Expected values
+    real y_new[R, T, 7];  // Replicate data
+    vector[7] p[R, T];
     matrix[T, 7] E[R];
     matrix[T, 7] E_new[R];
     matrix[R, 7] ik_p;
-    vector[7] num_obs_site;
-
-    for (i in 1:R) {                         // Loop over R sites (95)
-      for (k in 1:(first[i] - 1)) {
-        N[i, k] <- 0;
-        ik_p[i, k] <- 0;
-        E[i, 1:T, k] <- rep_vector(0, T);
-        E_new[i, 1:T, k] <- rep_vector(0, T);
-      }
-      for (k in (last[i] + 1):7) {
-        N[i, k] <- 0;
-        ik_p[i, k] <- 0;
-        E[i, 1:T, k] <- rep_vector(0, T);
-        E_new[i, 1:T, k] <- rep_vector(0, T);
-      }
-      for (k in first[i]:last[i]) {          // Loop over days (7)
-        vector[K+1] pr;
 
-        pr <- softmax(lp[i, k]);
-        N[i, k] <- categorical_rng(pr) - 1;
-        for (j in 1:T) {
-          p[i, j, k] <- inv_logit(logit_p[i, j, k]);
+    // Initialize N and ik_p
+    N = rep_array(0, R, 7);
+    ik_p = rep_matrix(0, R, 7);
+    for (i in 1:R) {
+      // Initialize E and E_new
+      E[i] = rep_matrix(0, T, 7);
```

Comment:
If there are efficiency issues, this will be more efficient if you set

```
E[i] = E[i - 1];
```

for cases where `i > 1`